internal/pkg/agent/application/actions/handlers/handler_action_upgrade.go

```diff
@@ -30,14 +35,90 @@ func NewUpgrade(log *logger.Logger, coord *coordinator.Coordinator) *Upgrade {
 	}
 }
 
-// Handle handles UPGRADE action.
-func (h *Upgrade) Handle(ctx context.Context, a fleetapi.Action, _ acker.Acker) error {
+// Handle handles UPGRADE action.  Returns immediately and the actual
+// upgrade happens asynchronously.  This allows for downloads to
+// happen without blocking updates.  If multiple upgrades are sent
+// then we ack them all if there is an error, but only the first actually executes.
+// If successful, reboot does ACK and check-in.
+func (h *Upgrade) Handle(ctx context.Context, a fleetapi.Action, ack acker.Acker) error {
```

Comment:
Overall this has gotten complex, so we should really add some unit tests for this.